frame/executive/src/lib.rs

```diff
@@ -101,6 +98,37 @@ pub type CheckedOf<E, C> = <E as Checkable<C>>::Checked;
 pub type CallOf<E, C> = <CheckedOf<E, C> as Applyable>::Call;
 pub type OriginOf<E, C> = <CallOf<E, C> as Dispatchable>::Origin;
 
+/// Stores when the latest runtime upgrade happened.
+///
+/// This is stored under the `well_know_keys::LAST_RUNTIME_UPGRADE` key.
+#[derive(Default, Encode, Decode)]
+struct LastRuntimeUpgrade {
+	spec_version: u32,
+	impl_version: u32,
+}
+
+impl LastRuntimeUpgrade {
+	/// Checks if in respect to `current` version there was a runtime upgrade.
+	fn was_upgraded(&self, current: &sp_version::RuntimeVersion) -> bool {
+		if current.spec_version > self.spec_version {
+			true
+		} else if current.spec_version == self.spec_version {
+			current.impl_version > self.impl_version
```

Comment:
Should we run the runtime upgrade logic if impl version is bumped only?

Reply:
Good question. And probably you are right.

Reply:
I thought again about this and I think it is okay to run runtime upgrade logic as well on `impl_version` bump.

I think there are multiple factors:
- Don't add an extra footgun for people that may forgot to increment the `spec_version`.
- With the old `RuntimeUpgraded` flag version, it would also execute the upgrade logic.
- You probably will not issue a runtime upgrade that just contains a bump of the `impl_version`. That is probably just for node releases. (Goes into the direction of point 1)

Reply:
I can see you argument, but also think about a person who forgets to impl spec when they make a breaking change. Their network will fork. So we dont really protect them much here I think.

Reply:
Agree. And as long as we have the semantic difference i think we should be using them.

Reply:
To mitigate the potential foot-gun, we should ensure that `set_code` verifies that spec_version is different.

Reply:
Addressed.